thirdparty/patchmatchnet/models/net.py

```diff
@@ -0,0 +1,373 @@
+                    stage=i + 1,
+                    evaluate_neighbors=evaluate_neighbors[i],
+                )
+            setattr(self, f"patchmatch_{i+1}", patchmatch)
+
+        self.upsample_net = Refinement()
+
+    def forward(
+        self,
+        imgs: Dict[str, torch.Tensor],
+        proj_matrices: Dict[str, torch.Tensor],
+        depth_min: torch.Tensor,
+        depth_max: torch.Tensor,
+    ) -> Dict[str, Any]:
+        """Forward method for PatchMatchNet
+
+        Args:
+            imgs: different stages of images (B, 3, H, W) stored in the dictionary
+            proj_matrics: different stages of camera projection matrices (B, 4, 4) stored in the dictionary
+            depth_min: minimum virtual depth (B, )
+            depth_max: maximum virtual depth (B, )
+
+        Returns:
+            output dictionary of PatchMatchNet, containing refined depthmap, depth patchmatch
+                and photometric_confidence.
+        """
+        imgs_0 = torch.unbind(imgs["stage_0"], 1)
+        imgs_1 = torch.unbind(imgs["stage_1"], 1)
+        imgs_2 = torch.unbind(imgs["stage_2"], 1)
+        imgs_3 = torch.unbind(imgs["stage_3"], 1)
+        del imgs
+
+        self.imgs_0_ref = imgs_0[0]
+        self.imgs_1_ref = imgs_1[0]
+        self.imgs_2_ref = imgs_2[0]
+        self.imgs_3_ref = imgs_3[0]
+        del imgs_1, imgs_2, imgs_3
+
+        self.proj_matrices_0 = torch.unbind(proj_matrices["stage_0"].float(), 1)
+        self.proj_matrices_1 = torch.unbind(proj_matrices["stage_1"].float(), 1)
+        self.proj_matrices_2 = torch.unbind(proj_matrices["stage_2"].float(), 1)
+        self.proj_matrices_3 = torch.unbind(proj_matrices["stage_3"].float(), 1)
+        del proj_matrices
+
+        assert len(imgs_0) == len(self.proj_matrices_0), "Different number of images and projection matrices"
+
+        # step 1. Multi-scale feature extraction
+        features = []
+        for img in imgs_0:
+            output_feature = self.feature(img)
+            features.append(output_feature)
+        del imgs_0
+        ref_feature, src_features = features[0], features[1:]
+
+        depth_min = depth_min.float()
+        depth_max = depth_max.float()
+
+        # step 2. Learning-based patchmatch
+        depth = None
+        view_weights = None
+        depth_patchmatch = {}
+        refined_depth = {}
+
+        for l in reversed(range(1, self.stages)):
+            src_features_l = [src_fea[f"stage_{l}"] for src_fea in src_features]
+            projs_l = getattr(self, f"proj_matrices_{l}")
+            ref_proj, src_projs = projs_l[0], projs_l[1:]
+
+            if l > 1:
+                depth, _, view_weights = getattr(self, f"patchmatch_{l}")(
+                    ref_feature=ref_feature[f"stage_{l}"],
+                    src_features=src_features_l,
+                    ref_proj=ref_proj,
+                    src_projs=src_projs,
+                    depth_min=depth_min,

```

Comment:
`depth_min` here is supposed to be an int, but we say it is a tensor of size (B,)? Can you double check these by stepping through it @codyly ?

Reply:
All `depth_min` /  `depth_max` should  be `torch.Tensor` with shape (B,). I missed some in `patchmatch.py`